src/csharp/Microsoft.Spark/Interop/Ipc/CallbackServer.cs

```diff
@@ -0,0 +1,256 @@
+
+        /// <summary>
+        /// Keeps track of all <see cref="CallbackConnection"/> objects identified by its
+        /// <see cref="CallbackConnection.ConnectionId"/>. The main thread creates a
+        /// <see cref="CallbackConnection"/> each time it receives a new socket connection
+        /// from the JVM side and inserts it into <see cref="_connections"/>. Each worker
+        /// thread calls <see cref="CallbackConnection.Run"/> and removes the connection
+        /// once this call is finished. <see cref="CallbackConnection.Run"/> will not return
+        /// unless the <see cref="CallbackConnection"/> needs to be closed.
+        /// Also, <see cref="_connections"/> is used to bound the number of worker threads
+        /// since it gives you the total number of active <see cref="CallbackConnection"/>s.
+        /// </summary>
+        private readonly ConcurrentDictionary<long, CallbackConnection> _connections =
+            new ConcurrentDictionary<long, CallbackConnection>();
+
+        /// <summary>
+        /// Each worker thread picks up a CallbackConnection from _waitingConnections
+        /// and runs it.
+        /// </summary>
+        private readonly BlockingCollection<CallbackConnection> _waitingConnections =
+            new BlockingCollection<CallbackConnection>();
+
+        /// <summary>
+        /// A <see cref="CancellationTokenSource"/> used to notify threads that operations
+        /// should be canceled.
+        /// </summary>
+        private readonly CancellationTokenSource _tokenSource = new CancellationTokenSource();
+
+        /// <summary>
+        /// Counter used to generate a unique id when registering a <see cref="ICallbackHandler"/>.
+        /// </summary>
+        private int _callbackCounter = 0;
+
+        private bool _isRunning = false;
+
+        private ISocketWrapper _listener;
+
+        internal int CurrentNumConnections => _connections.Count;
+
+        internal CallbackServer(IJvmBridge jvm, bool run = true)
+        {
+            AppDomain.CurrentDomain.ProcessExit += (s, e) => Shutdown();
+            _jvm = jvm;
+
+            if (run)
+            {
+                Run();
+            }
+        }
+
+        /// <summary>
+        /// Produce a unique id and register a <see cref="ICallbackHandler"/> with it.
+        /// </summary>
+        /// <param name="callbackHandler">The handler to register.</param>
+        /// <returns>A unique id associated with the handler.</returns>
+        internal int RegisterCallback(ICallbackHandler callbackHandler)
+        {
+            int callbackId = Interlocked.Increment(ref _callbackCounter);
+            _callbackHandlers[callbackId] = callbackHandler;
+
+            return callbackId;
+        }
+
+        /// <summary>
+        /// Runs the callback server.
+        /// </summary>
+        internal void Run()
+        {
+            _listener = SocketFactory.CreateSocket();
```

Comment:
can be problematic if `_isRunning == true`?

Reply:
maybe should just make `Run()` private.

Reply:
that's also fine with me.